### DISCUSSION TITLE
[Feature] Smart merge of duplicated assets

### DISCUSSION THREAD

**Discussion by metheoryt** [immich-app/immich#10846] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

Currently, by default, duplicated assets are deduplicated by leaving the biggest file.

There is a big room for improvement, especially when using bulk deduplicate function, because right now there is no way to determine, whether bulk deduplicate is leaving the right file as original for each group.

- Asset format isn't respected when comparing by asset size - original HEIC image can be smaller than its JPEG duplicate;
- EXIF dates can be utilized to understand, which asset was created earlier - usually, oldest one is the original;
- Speaking about EXIF - this data can be merged from duplicates into original, as well as Immich-managed metadata, such as "To which albums do duplicates belong". For example, merging oldest "media created date" into biggest asset can be useful for assets, that were saved multiple times, each time with better quality (imagine saving compressed photos from messenger and then saving their better versions that lost their creation date - right now deduplication allows leave either quality file, or file with quality metadata).

In general, the goal of smart merge is to preserve all the best from duplicate group - highest quality content, maximum metadata, oldest dates.

### Platform

- [X] Server
- [x] Web
- [ ] Mobile

**Comment by mertalev** [comment]
> Asset format isn't respected when comparing by asset size - original HEIC image can be smaller than its JPEG duplicate

We plan to introduce an image quality assessment model that can detect the image with the least compression artifacts, least blurriness, etc. It can be reused in other cases as well, such as when generating person thumbnails and allowing users to search explicitly for low quality images.

> EXIF dates can be utilized to understand, which asset was created earlier - usually, oldest one is the original

This is possible, but EXIF data is unfortunately not always reliable; there are cases where the date will be set to 1970, 1900, some other incorrect date because the camera was never configured, etc.

> Speaking about EXIF - this data can be merged from duplicates into original, as well as Immich-managed metadata, such as "To which albums do duplicates belong". For example, merging oldest "media created date" into biggest asset can be useful for assets, that were saved multiple times, each time with better quality (imagine saving compressed photos from messenger and then saving their better versions that lost their creation date - right now deduplication allows leave either quality file, or file with quality metadata

Adding kept assets to the albums of duplicate assets is planned. I imagine merging EXIF metadata would be non-trivial since metadata for a duplicate may not be appropriate for the original, and again there may be mistakes in the metadata itself. If we handled it automatically, it would inevitably lead to mistakes that are hard to notice until later. If added, I think this would be something you configure rather than us opaquely deciding which metadata field to use from which asset.

**Reply by metheoryt** [reply]
True, bulk deduplication can never be trusted fully, and I would reference iOS Photos app deduplication experience here. 
It shows all duplicate groups, allows to manually select those that can be merged in bulk, or select all groups at once. 

Within single duplicates group it shows how exactly the group will be merged - it shows source files and their metadata, and resulting file and its metadata - so I can see what the resulting asset will take from any of duplicates in the group. I'm not sure which metadata is being merged by the app, but this should be at least creation date. Also, it seems like a quick win to me to fill missing exif data.

**Reply by githubsean** [reply]
> Adding kept assets to the albums of duplicate assets is planned.

Even if merging EXIF data is complicated (please try though), keeping Album information would be great - this is "Immich added" information. For EXIF data, I am going to think that the highest quality version (by whatever metric) will likely be the original and probably have the most reliable EXIF data and you can just discard the other ones.

**Comment by vw-kombi** [comment]
From my experience here as a user of only 1 week, I would have liked a tickbox to prioritise a smaller HEIC over a larger JPG file in the dupe checker.  We are a family of iphones and in the old days we connected them to the computer, and imported the HEIC files and they converted to JPG's as part of the backup process we used.  

Now in immich we have a JPG and a HEIC of many thousands of photos after an immich-go import of the filesystem on the PC and the icloudPD backups from the phones.  

After this latest version of immich, I let it mass delete the dupes, and hence it mass deleted all the HEIC's (as they are smaller that the corresponding jpg).  Then I run immich-go again to select my icloudPD folder for the iphones and a large number of the HEIC's are back again - which would not have happened if the dupe detector prioritised HEIC.  HOWEVER, I did this process again (run mass dedupe, then added icloudPD again) and this time no new dupes were created....... not sure what happened there - may have been an immich-go issue.  I have repeated this import over and over from each set of icloudPD folders (from each iphone), and no more dupes have been added.

I would like my workflow to be thus :
1 - Iphones use Icloud, for device backups, and saving photos to icloud when in wifi
2 - IcloudPD runs automatically every day to download any new icloud photos for each iphone.
3 - immich-go runs every week (or on demand) to update immich photos from icloudPD with any new photos taken on iphones 

I believe for my planned workflow, immich will work fine for me if there was a setting in the dupe detector to override HEIC over JPG.  Once thats there, I will re-setup all the immich config from scratch and re-import all sources, then dedupe again.  Hence keeping the HEIC duplicate instead of the JPG.

**Comment by ldivl** [comment]
I really appreciate the dedupe feature! However, I would find it even more useful if there was an option to hide duplicates instead of deleting them. This would allow users to review duplicates later without permanently losing any data. Is this something that could be considered for future updates? Thanks for your hard work on this project!

**Reply by mertalev** [reply]
We plan to let you stack assets instead of trashing them.

**Reply by AngelaDMerkel** [reply]
Maybe there could be some other modality for accessing deduplicated images? Stacks for visually similar images are very different from stacks of duplicate images which might have been made through user error and which might be several years apart in the timeline. A "view duplicates" panel would be useful in this case, I feel.

**Comment by petersuba** [comment]
To add, I think there is also the use case where we have pictures taken in sequence - similar to live photos, but manually taking pictures in sequence, for example of a group of people to then pick the best (i.e. one where no one has closed eyes, pulling faces etc). 
This will typically result in these identified as duplicates (correctly).
Now, the ideal solution of this to de-duplicate is to replace the series with a "smart blend" i.e. pick the best base image but replace the odd face/eyes which as not optimal in that photo (e.g. in a photo of 20 people if 1 person is blinking and 1 looks to the side, replace their face from a nearby picture where they are). This is obviously quite advanced and not MVP ... but as we ideally progress through to this, there could be implemented in stages such as
1. identify this scenario of duplications (i.e. detected duplicates that are taken only second apart from each other)
2. show the pictures in this scenario in chronological order to help cycle through and pick accordingly
3. instead of just "remove/keep" add markers for "remove/keep as a stack/presentation picture"
4. in the "keep as stack", have the ability to mark inside the photo the sections that are preferred to be kept or discarded (i.e. the best face of someone in a sequence, the one which we definitely don't want to keep". This could be combined with or using the face detection results (i.e. instead of drawing manual squares, use the detected regions and mark them)
5. Full implementation where the additional blended photo is generated

**Comment by spawnpeekboi** [comment]
I think that merging metadata will be a really good idea
as sometimes we manually push our data to backup to wherever it is, then we maybe just simply pull all data and dumb it into immich CLI for upload. However, sometimes the metadata like creation date get mess up. For example, iCloud take-out will alter the date of creation to the date of the file being download. Having the choice of choosing which meta data to merge with while purging the duplicated one will be a good idea. There can be a checkbox list which put both metadata side by side, defaulted choosing metadata of the one having higher quality using the image quality assessment model mentioned above, and override the defaulted choice for some option, like choosing the older creation date, choosing having geolocation then not having, choosing having camera info then not having.

![image](https://github.com/user-attachments/assets/b32d7d99-938e-4656-92df-80ce79023b71)
(sorry for drawing this in paint, me dun know UI design)

**Reply by AngelaDMerkel** [reply]
This should be the standard unless the older file is 1970, because that is most likely to be an error

**Comment by dRuEFFECT** [comment]
i desperately need this functionality to clean up decades of duplicated images with a lot of messed up dates. keeping the largest file doesn't reliably keep the best data. i wish i could upvote this more.

**Comment by bo0tzz** [comment]
#10262
